homeassistant/helpers/script.py

```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
         if not unsub:
             return
 
-        if not self._stop.is_set():
-            self._changed()
+        self._changed()
         try:
-            timeout = self._action[CONF_TIMEOUT].total_seconds()
+            delay = self._action[CONF_TIMEOUT].total_seconds()
         except KeyError:
-            timeout = None
+            delay = None
         done = asyncio.Event()
         try:
-            await asyncio.wait_for(
-                asyncio.wait(
+            async with timeout(delay):
+                _, pending = await asyncio.wait(
                     {self._stop.wait(), done.wait()},
                     return_when=asyncio.FIRST_COMPLETED,
-                ),
-                timeout,
-            )
+                )
+            for pending_task in pending:
+                pending_task.cancel()
         except asyncio.TimeoutError:
             if not self._action.get(CONF_CONTINUE, True):
                 self._log(_TIMEOUT_MSG)
                 raise _StopScript
         finally:
             unsub()
 
+    async def _async_call_service_step(self):
+        """Call the service specified in the action."""
+        domain, service, service_data = self._prep_call_service_step()
+        blocking = service_data.get(ATTR_BLOCKING, True)
+
+        # If this might start a fully blocking script call then disable the call
+        # timeout. Otherwise use the normal service call limit.
+        if domain == "script" and (
+            service == SERVICE_TURN_ON and blocking or service != SERVICE_TURN_OFF
+        ):
+            limit = None
```

Comment:
This can be replaced by 

```suggestion
        if domain == "script" and service != SERVICE_TURN_OFF:
            limit = None
```

Reply:
✔️